extensions/script-libraries/httpd/httpd.livecodescript

```diff
@@ -0,0 +1,477 @@
+/*
+Copyright (C) 2017 LiveCode Ltd.
+
+This file is part of LiveCode.
+
+LiveCode is free software; you can redistribute it and/or modify it under
+the terms of the GNU General Public License v3 as published by the Free
+Software Foundation.
+
+LiveCode is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
+
+on revLoadLibrary
+   if the target is not me then
+      pass revLoadLibrary
+   end if
+   
+   insert the script of me into back
+   
+   if the environment contains "development" then
+      set the _ideoverride of me to true
+      revSBAddAvailableLibrary the short name of me, "HTTP Server"
+   end if
+end revLoadLibrary
+
+on revUnloadLibrary
+   if the target is not me then
+      pass revUnloadLibrary
+   end if
+   
+   remove the script of me from back
+end revUnloadLibrary
+
+/**
+
+Handle HTTP requests in your application
+
+*/
+
+local sServers -- an array of server info keyed on server port
+local sRequests -- an array of request info keyed on socket id
+
+
+/**
+
+Start a HTTP server
+
+Example:
+on mouseUp
+   httpdStart "NewRequest", 12345, "My Server"
+   launch url ("http://localhost:" & it)
+end mouseUp
+
+on NewRequest pSocketID, pRequest
+   httpResponse pSocketID, 200
+end NewRequest
+
+Description:
+Start a HTTP server. Once started any HTTP requests to the server will
+result in a callback to the caller with the message specified with pCallback.
+
+If the callback will have two parameters. The socket ID of the request. And an
+array describing the request. The array has the following keys:
+
+- "method": The HTTP request method
+- "resource": The request resource path e.g /images/foo.gif
+- "parameters": For GET requests with parameters or application/x-www-form-urlencoded encoded
+this key will be an array of parameters
+- "headers": An array of request headers
+- "content": The request body (empty for application/x-www-form-urlencoded)
+
+Parameters:
+
+pCallback (string):
+A handler name to be sent to the caller when a new HTTP request
+is handled by the server.
+
+pPort (optional string):
+The port to accept HTTP requests on. If empty 8080 will be tried
+and if the command fails the port will increment until a port
+is opened.
+
+pServerName (optional string):
+Optional server name defaults to `LiveCode`
+
+The result:
+An error string if an error occurred
+
+It:
+The port the server was started on.
+
+*/
+command httpdStart pCallback, pPort, pServerName
+   if pServerName is empty then
+      put "LiveCode" into pServerName
+   end if
+   
+   local tPort
+   if pPort is empty then
+      put 8080 into tPort
+      
+      repeat
+         accept connections on port tPort with message "__NewRequest"
```

Comment:
This loop isn't increasing the port at the moment :)

I suggest we perhaps use the 'newer' ephemeral port range if there is no port specified. First use 8080 (as that seems a reasonable default) - otherwise perhaps start with port number 41952 (see http://www.ncftp.com/ncftpd/doc/misc/ephemeral_ports.html)...

Even better, we should add a library at some point which gives us the platforms 'ephemeral port' / 'private port' range :)